internal/compiler/llr/optim_passes/count_property_use.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+//! Passes that fills the Property::use_count
+//!
+//! This pass assume that use_count of all properties is
+
+use crate::llr::{EvaluationContext, Expression, ParentCtx, PropertyReference, PublicComponent};
+
+pub fn count_property_use(root: &PublicComponent) {
+    // Visit the root property that are used.
+    // 1. the public properties
+    let root_ctx = EvaluationContext::new_sub_component(root, &root.item_tree.root, (), None);
+    for (_, pr) in root.public_properties.values() {
+        visit_property(pr, &root_ctx);
+    }
+    for g in root.globals.iter().filter(|g| g.exported) {
+        let ctx = EvaluationContext::new_global(root, g, ());
+        for (_, pr) in g.public_properties.values() {
+            visit_property(pr, &ctx);
+        }
+    }
+
+    root.for_each_sub_components(&mut |sc, ctx| {
+        // 2. the native items and bindings of used properties
+        for (pr, expr) in &sc.property_init {
+            let c = expr.use_count.get();
+            if c > 0 {
+                continue;
+            }
+            match pr {
+                PropertyReference::Local { sub_component_path, property_index } => {
+                    let mut sc = sc;
+                    for i in sub_component_path {
+                        sc = &sc.sub_components[*i].ty;
+                    }
+                    if sc.properties[*property_index].use_count.get() == 0 {
```

Comment:
Is it guaranteed that `use_count` is up-to-date at this point? (i.e. the final value) What if the property's use count is increased later due to another use-site?

Reply:
Not the final value, but good enough for what we care about here: is the value read by the sub component itself.
For local properties, we will go recursively to find them.